### PR TITLE
#46 Basic Excel File Import 

### PR DESCRIPTION
**Pull request by rahuang**
#46 #55 #56 #57 #59 
- gems
  - roo (general gem for excel and csv upload/parsing)
  - roo-xls (gem used for excel 2003 file extension .xls)
  - activerecord-session_store (Generally the excel file is stored in the session which is stored in the cookies. However, cookies only have a 4kb of storage, which is not enough for the excel file. So this gem is needed to help with the process of storing session data in a sessions table in the database instead of the cookie. The cookie will only store the key that is used to find the data in teh sessions table.)
- root path for excel upload
- only accepts .xls and xlsx
- IMPORTS ALL MODELS!!!
- Attached is the test excel file that I used
- Fixed 2 bugs
  - Do begin rescue clause around all parsing in case of errors with incorrect sheet names and column names
  - For models with foreign keys, added check if the excel cell is nil or out of bounds of foreign model and sets to nil for now (will be caught by the validation and throw a missing foreign key error)
- 100% TEST COVERAGE YEAH!!!!!!!
[competencies.xlsx](https://github.com/373-CMU-PDS/Competency_Assessment/files/873503/competencies.xlsx)

### CODE REVIEW

File: app/models/indicator.rb [hunk 1]
```diff
@@ -23,4 +23,28 @@ class Indicator < ActiveRecord::Base
...
+    indicators = []
+    indicators_hash.each_with_index do |i, index|
+      indicator = Indicator.new
+      # Used to set the foreign keys. 
+      # competency_id is always set to the first competency, since there's only 1 competency per spreadsheet
+      i[:competency_id] = competencies[0].nil? ? nil : competencies[0].id
+      # level_id should be the id of the level relative to the row number of levels
+      # the offset is because of 0 indexing plus the header in the excel file
+      lid = i[:level_id]
+      i[:level_id] = (lid.nil? || levels[lid - 2].nil?) ? nil : levels[lid - 2].id
```
Comment (ppuri96) [inline app/models/indicator.rb]: Add a comment explaining this logic and why it is necessary.

File: app/models/indicator.rb [hunk 2]
```diff
@@ -23,4 +23,28 @@ class Indicator < ActiveRecord::Base
...
+  # Methods
+  def self.parse(spreadsheet, competencies, levels)
+    indicators_sheet = spreadsheet.sheet("Indicators")
+    indicators_hash = 
+      indicators_sheet.parse(level_id: "Level_ID", description: "Description")
+
+    indicators = []
+    indicators_hash.each_with_index do |i, index|
+      indicator = Indicator.new
+      # Used to set the foreign keys.
```
Comment (ppuri96) [inline app/models/indicator.rb]: Space out comment blocks to make the code more readable.

File: app/controllers/imports_controller.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+      end
+    end
+    # If all are valid, then save them and return true
+    models_list.each do |models|
+      models.each(&:save!)
+    end
+    return true
+  end
+
+  # This method is called if an error exists in the list of models.
```
Comment (tchambers3) [inline app/controllers/imports_controller.rb]: Maybe add into the comment what the different errors could be. Blank row, index out of bounds etc. Probably not completely necessary but might be nice for future teams.

**Review by tchambers3 — APPROVED** [review]
It looks good to me. The only thing I thought could be added was a comment about what errors could make the upload fail but it's not 100% necessary.

**Pull request merged**